This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [ ]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = 'FyxQ2HRwPiS6KKcxnL-v'

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [3]:
# First, import the relevant modules
import quandl
import requests
import pandas as pd
from pandas.io.json import json_normalize
import json

In [4]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
mydata = quandl.get("FSE/AFX_X", authtoken="FyxQ2HRwPiS6KKcxnL-v", start_date="2017-1-1", end_date="2017-12-31")

In [5]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
mydata.head()

Open   High    Low  Close  Change  Traded Volume   Turnover  \
Date                                                                       
2017-01-02  34.99  35.94  34.99  35.80     NaN        44700.0  1590561.0   
2017-01-03  35.90  35.93  35.34  35.48     NaN        70618.0  2515473.0   
2017-01-04  35.48  35.51  34.75  35.19     NaN        54408.0  1906810.0   
2017-01-05  35.02  35.20  34.73  35.06     NaN        48412.0  1692326.0   
2017-01-06  34.91  35.21  34.91  35.04     NaN        27507.0   964046.0   

           Last Price of the Day Daily Traded Units Daily Turnover  
Date                                                                
2017-01-02                  None               None           None  
2017-01-03                  None               None           None  
2017-01-04                  None               None           None  
2017-01-05                  None               None           None  
2017-01-06                  None               None           None

In [6]:
mydata.sort_values(by = ['Open'], ascending = True) #highest was 53.11, lowest was 34.00

Open    High     Low  Close  Change  Traded Volume   Turnover  \
Date                                                                         
2017-01-24  34.00  34.350  33.850  34.22     NaN        48797.0  1666086.0   
2017-01-23  34.04  34.120  33.620  34.06     NaN        55333.0  1877957.0   
2017-01-25  34.42  34.860  34.030  34.83     NaN        56240.0  1947147.0   
2017-01-20  34.54  34.590  34.050  34.17     NaN        80246.0  2743474.0   
2017-02-01  34.75  36.000  34.750  35.94     NaN        85137.0  3038172.0   
...           ...     ...     ...    ...     ...            ...        ...   
2017-12-19  52.73  52.730  51.070  51.66     NaN       137313.0  7102361.0   
2017-12-14  53.11  53.540  52.150  52.67     NaN       132981.0  7016953.0   
2017-04-14    NaN  42.480  41.985  42.20     NaN        88416.0  3734717.0   
2017-04-17    NaN  42.480  41.985  42.20     NaN        88416.0  3734717.0   
2017-05-01    NaN  42.245  41.655  41.72   -0.44        86348.0  3606589.0   

           Last Price of the Day Daily Traded Units Daily Turnover  
Date                                                                
2017-01-24                  None               None           None  
2017-01-23                  None               None           None  
2017-01-25                  None               None           None  
2017-01-20                  None               None           None  
2017-02-01                  None               None           None  
...                          ...                ...            ...  
2017-12-19                  None               None           None  
2017-12-14                  None               None           None  
2017-04-14                  None               None           None  
2017-04-17                  None               None           None  
2017-05-01                  None               None           None  

[255 rows x 10 columns]

In [110]:
#mydata[['High', 'Low']].pct_change(axis = 'columns').max()*100
highest_change = mydata['High'] - mydata['Low']
max(highest_change)

2.8100000000000023

In [120]:
mydata[['Close']].pct_change(periods = 2).max()*100

Close    6.289655
dtype: float64

In [121]:
mydata[['Traded Volume']].mean()

Traded Volume    89124.337255
dtype: float64

In [122]:
mydata[['Traded Volume']].median()

Traded Volume    76286.0
dtype: float64

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [21]:
r = requests.get('https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=FyxQ2HRwPiS6KKcxnL-v')

In [ ]:
https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=FyxQ2HRwPiS6KKcxnL-v

In [23]:
r.json().keys()

dict_keys(['dataset'])

In [24]:
r.json()['dataset'].keys()

dict_keys(['id', 'dataset_code', 'database_code', 'name', 'description', 'refreshed_at', 'newest_available_date', 'oldest_available_date', 'column_names', 'frequency', 'type', 'premium', 'limit', 'transform', 'column_index', 'start_date', 'end_date', 'data', 'collapse', 'order', 'database_id'])

In [25]:
lst = r.json()['dataset']['data']

In [26]:
lst

[['2017-12-29',
  51.76,
  51.94,
  51.45,
  51.76,
  None,
  34640.0,
  1792304.0,
  None,
  None,
  None],
 ['2017-12-28',
  51.65,
  51.82,
  51.43,
  51.6,
  None,
  40660.0,
  2099024.0,
  None,
  None,
  None],
 ['2017-12-27',
  51.45,
  51.89,
  50.76,
  51.82,
  None,
  57452.0,
  2957018.0,
  None,
  None,
  None],
 ['2017-12-22',
  51.05,
  51.5,
  50.92,
  51.32,
  None,
  71165.0,
  3641949.0,
  None,
  None,
  None],
 ['2017-12-21',
  51.16,
  51.52,
  50.9,
  51.4,
  None,
  120649.0,
  6179433.0,
  None,
  None,
  None],
 ['2017-12-20',
  51.88,
  52.04,
  51.2,
  51.27,
  None,
  50587.0,
  2610258.0,
  None,
  None,
  None],
 ['2017-12-19',
  52.73,
  52.73,
  51.07,
  51.66,
  None,
  137313.0,
  7102361.0,
  None,
  None,
  None],
 ['2017-12-18',
  52.37,
  52.75,
  51.61,
  52.62,
  None,
  129733.0,
  6770499.0,
  None,
  None,
  None],
 ['2017-12-15',
  52.7,
  52.7,
  51.64,
  52.01,
  None,
  204080.0,
  10596319.0,
  None,
  None,
  None],
 ['2017-12-14',
  53.

In [39]:
keys = ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last', 'Price of the Day', 'Daily Traded Units', 'Daily Turnover']

In [41]:
list_dict = {k: v for k, *v in zip(keys, *lst)}

In [59]:
list_dict

{'Date': ['2017-12-29',
  '2017-12-28',
  '2017-12-27',
  '2017-12-22',
  '2017-12-21',
  '2017-12-20',
  '2017-12-19',
  '2017-12-18',
  '2017-12-15',
  '2017-12-14',
  '2017-12-13',
  '2017-12-12',
  '2017-12-11',
  '2017-12-08',
  '2017-12-07',
  '2017-12-06',
  '2017-12-05',
  '2017-12-04',
  '2017-12-01',
  '2017-11-30',
  '2017-11-29',
  '2017-11-28',
  '2017-11-27',
  '2017-11-24',
  '2017-11-23',
  '2017-11-22',
  '2017-11-21',
  '2017-11-20',
  '2017-11-17',
  '2017-11-16',
  '2017-11-15',
  '2017-11-14',
  '2017-11-13',
  '2017-11-10',
  '2017-11-09',
  '2017-11-08',
  '2017-11-07',
  '2017-11-06',
  '2017-11-03',
  '2017-11-02',
  '2017-11-01',
  '2017-10-30',
  '2017-10-27',
  '2017-10-26',
  '2017-10-25',
  '2017-10-24',
  '2017-10-23',
  '2017-10-20',
  '2017-10-19',
  '2017-10-18',
  '2017-10-17',
  '2017-10-16',
  '2017-10-13',
  '2017-10-12',
  '2017-10-11',
  '2017-10-10',
  '2017-10-09',
  '2017-10-06',
  '2017-10-05',
  '2017-10-04',
  '2017-10-02',
  '2017-09-29',


In [47]:
for i in list_dict:
    print(i)

Date
Open
High
Low
Close
Change
Traded Volume
Turnover
Last
Price of the Day
Daily Traded Units


In [73]:
open_list = list_dict['Open']

In [79]:
import numpy as np

In [80]:
open_list_nparray = np.array(open_list, dtype=np.float64) #converting to np array

In [81]:
open_list_nparray

array([51.76, 51.65, 51.45, 51.05, 51.16, 51.88, 52.73, 52.37, 52.7 ,
       53.11, 52.64, 52.29, 52.28, 51.5 , 50.89, 50.8 , 51.21, 49.5 ,
       49.52, 48.64, 49.64, 49.09, 49.13, 49.11, 48.8 , 48.4 , 47.25,
       46.57, 47.03, 47.09, 47.98, 48.4 , 48.38, 47.3 , 47.65, 46.42,
       46.16, 45.81, 45.  , 45.88, 46.29, 46.53, 45.48, 45.2 , 45.01,
       45.16, 44.9 , 45.08, 45.72, 46.01, 45.8 , 45.61, 45.5 , 45.58,
       45.97, 45.64, 46.2 , 46.19, 46.01, 45.36, 44.51, 43.58, 42.  ,
       42.35, 42.3 , 42.3 , 41.48, 42.29, 42.54, 42.65, 42.5 , 42.29,
       42.35, 42.49, 43.21, 42.81, 42.7 , 43.  , 42.66, 43.  , 42.38,
       42.16, 42.  , 42.  , 41.71, 42.11, 42.64, 42.72, 42.82, 42.46,
       42.42, 42.28, 41.88, 42.4 , 42.53, 42.12, 41.3 , 41.73, 43.5 ,
       44.9 , 45.85, 45.13, 45.34, 45.25, 45.24, 44.94, 45.26, 45.16,
       44.91, 44.7 , 45.31, 45.57, 45.74, 45.06, 45.5 , 45.6 , 45.07,
       44.67, 44.29, 44.94, 44.64, 44.79, 45.5 , 44.67, 45.83, 45.29,
       45.01, 45.73,

In [82]:
np.nanmax(open_list_nparray) #max open price

53.11

In [83]:
np.nanmin(open_list_nparray) #min open price

34.0

In [89]:
high_list = list_dict['High']

In [90]:
low_list = list_dict['Low']

In [91]:
high_list_nparray = np.array(high_list, dtype=np.float64)

In [92]:
low_list_nparray = np.array(low_list, dtype=np.float64)

In [93]:
difference = high_list_nparray - low_list_nparray

In [99]:
max(difference) #highest change between high and low in any one day

2.8100000000000023

In [113]:
close_list = list_dict['Close']

In [114]:
close_list_nparray = np.array(close_list, dtype=np.float64)

In [115]:
close_list_nparray

array([51.76, 51.6 , 51.82, 51.32, 51.4 , 51.27, 51.66, 52.62, 52.01,
       52.67, 53.09, 52.43, 52.14, 52.12, 51.47, 50.89, 51.25, 51.14,
       49.86, 49.7 , 48.75, 49.25, 49.2 , 49.11, 49.2 , 48.8 , 48.39,
       47.04, 46.84, 47.03, 47.05, 48.  , 48.34, 48.34, 47.21, 47.47,
       46.26, 45.99, 45.97, 45.27, 46.04, 45.76, 46.41, 45.41, 45.  ,
       44.85, 45.  , 44.87, 45.  , 45.77, 45.96, 45.55, 45.4 , 45.43,
       45.29, 45.84, 45.74, 46.  , 46.05, 46.11, 44.98, 44.17, 43.56,
       42.04, 42.37, 42.07, 42.06, 41.46, 41.99, 42.44, 42.27, 42.42,
       42.52, 42.45, 42.73, 42.85, 42.67, 42.77, 42.55, 42.62, 42.6 ,
       42.41, 41.9 , 41.94, 41.85, 41.91, 42.14, 42.69, 42.71, 42.71,
       42.26, 42.41, 42.5 , 42.05, 42.28, 42.3 , 41.94, 41.68, 41.81,
       44.37, 44.96, 45.07, 44.97, 45.56, 45.45, 45.3 , 44.97, 45.25,
       45.16, 44.82, 44.61, 45.44, 45.66, 45.57, 45.  , 45.6 , 45.53,
       44.95, 44.95, 44.2 , 44.7 , 44.53, 44.62, 45.19, 44.8 , 45.75,
       45.44, 45.2 ,

In [142]:
close_list_nparray[len(close_list_nparray)%2==0]

array([], shape=(0, 255), dtype=float64)

In [153]:
range(len(close_list_nparray))

range(0, 255)

In [152]:
for i in range(len(close_list_nparray)):
    if i % 2 == 0:
        print(close_list_nparray[i+2] - i)

51.82
49.4
47.66
46.01
45.09
42.14
39.47
37.25
33.86
30.75
29.200000000000003
27.200000000000003
24.39
20.840000000000003
19.049999999999997
18.340000000000003
15.21
12.259999999999998
9.969999999999999
8.04
6.409999999999997
3.0
1.0
-1.0
-2.039999999999999
-4.600000000000001
-6.710000000000001
-8.259999999999998
-9.950000000000003
-13.020000000000003
-16.439999999999998
-19.630000000000003
-21.939999999999998
-24.009999999999998
-25.729999999999997
-27.479999999999997
-29.270000000000003
-31.33
-33.45
-35.4
-38.1
-40.15
-41.86
-43.29
-45.74
-47.5
-49.72
-52.06
-54.19
-53.04
-55.03
-56.55
-59.03
-60.84
-63.39
-64.34
-67.0
-68.47
-71.05
-73.3
-75.38
-77.2
-78.56
-80.32
-80.78999999999999
-82.71000000000001
-84.63
-87.37
-89.67
-91.69
-93.72999999999999
-94.57
-97.47999999999999
-99.97999999999999
-102.7
-105.47
-108.16
-110.02000000000001
-110.69
-113.00999999999999
-116.72
-119.75999999999999
-121.5
-123.85
-126.28
-128.5
-130.19
-132.07
-134.0
-135.8
-137.8
-140.32
-142.95
-145.19
-14

IndexError: index 256 is out of bounds for axis 0 with size 255

In [121]:
tradedvolume = list_dict['Traded Volume']

In [122]:
tradedvolume_nparray = np.array(tradedvolume, dtype=np.float64)

In [123]:
np.mean(tradedvolume_nparray)

89124.33725490196

In [124]:
np.median(tradedvolume_nparray)

76286.0

Final Answers:

1. https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-01-01&end_date=2017-12-31&api_key=FyxQ2HRwPiS6KKcxnL-v
2. {k: v for k, *v in zip(keys, *lst)}
3. Highest Opening Price: 53.11 
   Lowest Opening Price: 34.0
4. Largest change in any one day (based on High and Low price): 2.8100000000000023
5. What was the largest change between any two days (based on Closing Price): 
6. Average daily trading volume during this year: 89124.33725490196
7. (Optional) Median trading volume during this year: 76286.0
    
    